### ISSUE TITLE
Istio 1.5 with istiod, missing telemetry /metrics port:42422, what is recommendation for prometheus?

### ISSUE THREAD

**Issue by kramvan1** [istio/istio#22777]
For users moving from 1.4 to 1.5 with istiod, what is the recommendation to replace the missing telemetry metrics port port 42422?   This was an easy prometheus scrape point to pick up the mesh metrics in 1.4.  In 1.5, what is the equivalent setup for metrics?

**Comment by kramvan1** [comment]
cc @linsun @howardjohn @rvennam

**Comment by howardjohn** [comment]
You can use the old mixer still, or set up telemetry v2 ( the default now): https://istio.io/docs/reference/config/telemetry/

**Comment by howardjohn** [comment]
are you using prometheus operator or just normal prometheus?

**Comment by kramvan1** [comment]
Would be using normal prometheus and telemetry v2.  But not sure on the setup here.

**Comment by howardjohn** [comment]
In the prometheus we ship we have this config:
```
    # Scrape config for envoy stats
    - job_name: 'envoy-stats'
      metrics_path: /stats/prometheus
      kubernetes_sd_configs:
      - role: pod

      relabel_configs:
      - source_labels: [__meta_kubernetes_pod_container_port_name]
        action: keep
        regex: '.*-envoy-prom'
      - source_labels: [__address__, __meta_kubernetes_pod_annotation_prometheus_io_port]
        action: replace
        regex: ([^:]+)(?::\d+)?;(\d+)
        replacement: $1:15090
        target_label: __address__
      - action: labeldrop
        regex: __meta_kubernetes_pod_label_(.+)
      - source_labels: [__meta_kubernetes_namespace]
        action: replace
        target_label: namespace
      - source_labels: [__meta_kubernetes_pod_name]
        action: replace
        target_label: pod_name
```

**Comment by kramvan1** [comment]
What extra prometheus scrape annotations are needed to pick that up for using a service like SysDig that support prometheus scraping?  is that the  `.Values.sidecarInjectorWebhook.injectedAnnotations`  `/stats/prometheus` with port `15090`?

**Comment by howardjohn** [comment]
So you are looking for adding a `prometheus.io/scrape` annotation basically? You can do it with injected annotations but it will break any existing annotations that are used for application metrics. https://github.com/istio/istio/pull/22318 fixes that which will be in 1.6 if it gets approved

**Comment by kramvan1** [comment]
Humm, don't think I want to break existing annotations.  Any chance a fix like that would be back ported?

**Comment by howardjohn** [comment]
Its a pretty large change, so its unlikely

**Comment by kramvan1** [comment]
Understand.  Is there a better approach that could be used for 1.5?

**Comment by howardjohn** [comment]
Is your requirement basically that you can only scrape things with `prometheus.io` annotations? If so I think your choices are to mangle app metrics or use mixer. Or deploy another prometheus and have it use the custom config above, then have your sysdig prometheus scrape that

**Comment by howardjohn** [comment]
No. Prometheus has scraping configuration, which controls which targets it scrapes. Istio ships with some configs. @kramvan1 uses a different set of configs, and because they are using some vendor's prometheus they cannot configure it. They config they use happens to be the industry standard; the istio config is non-standard.

The config istio ships works perfectly fine with v2 + user metrics. However, the config @kramvan1 has does not, until the change I mentioned lands.

Its not a strictly a "bug" in Istio, its more that @kramvan1 doesn't control their prometheus configuration (as is fairly common - thats why I am making the changes)

**Comment by linsun** [comment]
Thank you @howardjohn.  Would it be correct to say for 1.5, if mesh admin uses the default prometheus that we ship in istio, user will get app and proxy metrics.  However, if meshadmin wants to customize the prometheus config to scrape on certain annotations [1], we recommend mesh admin to switch to mixer telemetry in 1.5 before #21366 is resolved so that the annotations (app annotation by app owner + prometheus annotation by admin) can be merged properly.

[1] https://sysdig.com/blog/monitor-istio/

**Comment by howardjohn** [comment]
Its close. Its not if users "want to customize prometheus config", its more, "if users use a non-istio prometheus which they cannot customize". I also am very hestitant to recommend mixer, although I accidentally did it before. It would be better to just spin up your own prometheus for istio and have the other instance scrape that. See  https://preliminary.istio.io/docs/ops/best-practices/observability/#using-prometheus-for-production-scale-monitoring

**Comment by kramvan1** [comment]
For SysDig, need to config it like this for now in 1.5:
```
    prometheus:
      enabled: true
      process_filter:
         - include:
             process.name: envoy
             conf:
               port: 15090
               path: "/stats/prometheus"
```

**Comment by linsun** [comment]
Note: that is for sysdig agent config, or are you referring to config in IstioOperator for prometheus @kramvan1 ?

**Comment by kramvan1** [comment]
That is for the SysDig agent configmap.  The Istio Operator (IOP) cannot be used for this in 15  without also overriding the  customers annotations (fixed in 1.6).
 
Mark VanderwielIBM Cloud Platform - IBM Cloud Kubernetes ServiceRochester, MN==================================================================Internet Address: vanderwl@us.ibm.com
 
 
----- Original message -----From: Lin Sun <notifications@github.com>To: istio/istio <istio@noreply.github.com>Cc: Mark Vanderwiel <vanderwl@us.ibm.com>, Mention <mention@noreply.github.com>Subject: [EXTERNAL] Re: [istio/istio] Istio 1.5 with istiod, missing telemetry /metrics port:42422, what is recommendation for prometheus? (#22777)Date: Mon, Apr 13, 2020 9:08 PM 
 
Note: that is for sysdig agent config, or are you referring to config in IstioOperator for prometheus @kramvan1 ?

**Comment by kramvan1** [comment]
Better in 1.6, will close this.
